src/socketio.quiver.js

```diff
@@ -263,23 +300,59 @@ QuiverSocialProvider.prototype.login = function(loginOpts, continuation) {
             this.connectAsClient_.bind(this, friend, null)).
                 catch(onConnectionFailure.bind(this, friend)));
       }
-      Promise.all(connectionPromises).then(finishLogin);
+      return Promise.all(connectionPromises);
     }.bind(this);
 
     this.connectLoop_(this.configuration_.self.servers,
-        this.connectAsOwner_.bind(this)).then(connectToFriends, connectToFriends);
+        this.connectAsOwner_.bind(this)).then(function(results) {
+      if (results.succeeded.length > 0) {
+        // If at least one connection succeeded, then we do have a working
+        // network, so any connection failures are because the server is in fact
+        // failing or unreachable.  Remove it from the active list.
+        // It will then be replaced by a friend's server.
+        results.failed.forEach(this.removeServer_, this);
+      }
+      return connectToFriends();
+    }.bind(this)).then(function() {
+      var deficit = QuiverSocialProvider.MAX_CONNECTIONS_ - this.countOwnerConnections_();
+      if (deficit > 0) {
+        /** @type {!Object<string, QuiverSocialProvider.server_>} */
+        var unusedDefaultServers = {};
+            QuiverSocialProvider.DEFAULT_SERVERS_.forEach(function(server) {
```

Comment:
I'm kinda confused by the logic here.  In makeDefaultConfiguration_ you pick 2 random servers to be set to this.configuration_.self.servers.  But then here you aren't using those here - if I'm reading below correctly, it looks like you then take all of the DEFAULT_SERVERS (not just the 2 you picked), and for each one you haven't yet connected to as a client, you connect to them as an owner.

1. Shouldn't you instead be connecting to the this.configuration_.self.servers instead?
2. With the current code is it possible that unusedDefaultServers will end up being empty (if you've connected as a client to all of your DEFAULT_SERVERS)?